### PR TITLE
[SPARK-25307][SQL] ArraySort function may return an error in the code generation phase

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
Sorting array of booleans (not nullable) returns a compilation error in the code generation phase. Below is the compilation error :
```SQL
java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 51, Column 23: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 51, Column 23: No applicable constructor/method found for actual parameters "boolean[]"; candidates are: "public static void java.util.Arrays.sort(long[])", "public static void java.util.Arrays.sort(long[], int, int)", "public static void java.util.Arrays.sort(byte[], int, int)", "public static void java.util.Arrays.sort(float[])", "public static void java.util.Arrays.sort(float[], int, int)", "public static void java.util.Arrays.sort(char[])", "public static void java.util.Arrays.sort(char[], int, int)", "public static void java.util.Arrays.sort(short[], int, int)", "public static void java.util.Arrays.sort(short[])", "public static void java.util.Arrays.sort(byte[])", "public static void java.util.Arrays.sort(java.lang.Object[], int, int, java.util.Comparator)", "public static void java.util.Arrays.sort(java.lang.Object[], java.util.Comparator)", "public static void java.util.Arrays.sort(int[])", "public static void java.util.Arrays.sort(java.lang.Object[], int, int)", "public static void java.util.Arrays.sort(java.lang.Object[])", "public static void java.util.Arrays.sort(double[])", "public static void java.util.Arrays.sort(double[], int, int)", "public static void java.util.Arrays.sort(int[], int, int)"
	at com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:306)
	at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:293)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116)
	at com.google.common.util.concurrent.Uninterruptibles.getUninterruptibly(Uninterruptibles.java:135)
	at com.google.common.cache.LocalCache$Segment.getAndRecordStats(LocalCache.java:2410)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2380)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2342)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2257)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4000)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4004)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4874)
	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$.compile(CodeGenerator.scala:1305)

```

## How was this patch tested?
Added test in collectionExpressionSuite

### CODE REVIEW

**Comment by kiszk** [comment]
Good catch, LGTM

**Comment by dilipbiswal** [comment]
@kiszk Thanks. Can you please help look at [22315](https://github.com/apache/spark/pull/22315) as well.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by ueshin** [comment]
LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -1047,7 +1047,8 @@ trait ArraySortLike extends ExpectsInputTypes {
         s"int $c = ${ctx.genComp(elementType, s"(($jt) $o1)", s"(($jt) $o2)")};"
       }
       val nonNullPrimitiveAscendingSort =
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: `nonNullPrimitiveNumericAscendingSort` or `fastSortCode` preferred?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -1047,7 +1047,8 @@ trait ArraySortLike extends ExpectsInputTypes {
...
-        if (CodeGenerator.isPrimitiveType(elementType) && !containsNull) {
+        if (CodeGenerator.isPrimitiveType(elementType) && elementType != BooleanType
+          && !containsNull) {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: How about making this condition a separate method like `isFastSortSupported`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 1 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @maropu Lets keep the variable name the same .. but add a private def like you suggest in the following comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: will make a change

**Comment by dilipbiswal** [comment]
retest this please

**Comment by ueshin** [comment]
Thanks! merging to master.

**Comment by dilipbiswal** [comment]
@ueshin @kiszk @maropu  Thanks a lot.

**Comment by ueshin** [comment]
@dilipbiswal Do we need to backport this to 2.3? If so, could you submit a backport pr to branch-2.3 please? Thanks!

**Comment by dilipbiswal** [comment]
@ueshin Sure.

**Comment by dilipbiswal** [comment]
@ueshin Just verified in 2.3. This problem does not exist in 2.3. This is due to the fact that implementation of `nullSafeCodeGen` is different in 2.3 than in master. However, we are missing the test cases we added in these PRs in 2.3. Should we have the test cases checked in into the branch ? I am afraid that if we ever backported the pr that changed nullSafeCodeGen , we may introduce this bug. Please advise ..

**Comment by dilipbiswal** [comment]
@ueshin We don't have codegen enabled for this function in 2.3. So thankfully we don't have this problem :-)

**Comment by ueshin** [comment]
@dilipbiswal Thanks for confirming!

**Pull request closed without merging**